tests/upgrade-tests/utils.py

```diff
@@ -116,3 +121,48 @@ def all_units_ready(model):
     while not all_units_ready(model):
         assert_no_unit_errors(model)
         await asyncio.sleep(1)
+
+async def conjureup(namespace, bundle, channel, snap_channel, model):
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        cmd = 'charm pull --channel=%s cs:~%s/%s %s'
+        cmd %= channel, namespace, bundle, os.path.join(tmpdirname, bundle)
+        cmd = cmd.split()
+        process = await asyncio.create_subprocess_exec(*cmd)
+        await process.wait()
+        assert process.returncode == 0
+        shutil.copytree(
+            os.path.join('/snap/conjure-up/current/spells', bundle),
+            os.path.join(tmpdirname, 'spell')
+        )
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-01_get-kubectl'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-01_get-kubectl.yaml'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-02_cluster-info'))
+        os.remove(os.path.join(tmpdirname, 'spell', 'steps', 'step-02_cluster-info.yaml'))
```

Comment:
Would ignoring these steps if passwordless sudo is unavailable be an option? Perhaps print a warning - "kubectl and kubefed not installed on host"?

Bigger picture - feels a little weird that we're forcing this installation. Can it be prompted? If we can, then we can maybe skip the step if a tty isn't detected (in the headless case)?